insonmnia/gatekeeper/gatekeeper.go

```diff
@@ -0,0 +1,351 @@
+			zap.String("spentToday + value", spentToday.Add(spentToday, tx.Value).String()))
+		return false
+	}
+
+	return true
+}
+
+func (g *Gatekeeper) processUnpaidTransaction(ctx context.Context, tx *blockchain.GateTx) error {
+	if !g.underLimit(ctx, tx) {
+		g.logger.Debug("tx over keeper limit")
+		return fmt.Errorf("tx over keeper limit")
+	}
+
+	if !g.checkDelay(ctx, tx) {
+		g.logger.Debug("not cover delay check")
+		return fmt.Errorf("not cover delay check")
+	}
+
+	if !g.isNotPaid(ctx, tx) {
+		g.logger.Debug("transaction already paid")
+		return fmt.Errorf("transaction already paid")
+	}
+
+	return g.Payout(ctx, tx)
+}
+
+func (g *Gatekeeper) Payout(ctx context.Context, tx *blockchain.GateTx) error {
+	g.logger.Info("fix transaction",
+		zap.String("from", tx.From.String()),
+		zap.String("value", tx.Value.String()),
+		zap.String("tx number", tx.Number.String()))
+
+	if !g.isCommitted(ctx, tx) {
+		_, err := g.out.Payout(ctx, g.key, tx.From, tx.Value, tx.Number)
+		if err != nil {
+			g.logger.Debug("error while commit", zap.Error(err))
+			return err
+		}
+		g.logger.Info("transaction committed",
+			zap.String("from", tx.From.String()),
+			zap.String("value", tx.Value.String()),
+			zap.String("tx number", tx.Number.String()))
+
+		// sleeping for freezing time after committing
+		time.Sleep(g.freezingTime)
+	}
+
+	if !g.transactionCommittedByMe(ctx, tx) {
+		return fmt.Errorf("transaction commited by other gate")
+	}
+
+	_, err := g.out.Payout(ctx, g.key, tx.From, tx.Value, tx.Number)
+	if err != nil {
+		g.logger.Debug("error while payout", zap.Error(err))
+		return err
+	}
+	g.logger.Debug("transaction payouted",
+		zap.String("from", tx.From.String()),
+		zap.String("value", tx.Value.String()),
+		zap.String("tx number", tx.Number.String()))
+
+	return nil
+}
+
+func (g *Gatekeeper) findScummyTransactions(ctx context.Context, inTxs map[string]*blockchain.GateTx, outTxs map[string]*blockchain.GateTx) {
+	for k, tx := range outTxs {
+		_, ok := inTxs[k]
+		if !ok {
+			go g.freezeScummy(ctx, tx)
+		}
+	}
+}
+
+func (g *Gatekeeper) freezeScummy(ctx context.Context, tx *blockchain.GateTx) error {
+	txState, err := g.out.GetTransactionState(ctx, tx.From, tx.Value, tx.Number)
+	if err != nil {
+		return err
+	}
+
+	keeper, err := g.out.GetKeeper(ctx, txState.Keeper)
+	if err != nil {
+		return err
+	}
+
+	g.logger.Info("found new scum transaction",
+		zap.String("from", tx.From.String()),
+		zap.String("value", tx.Value.String()),
+		zap.String("tx number", tx.Number.String()),
+		zap.Uint64("block number", tx.BlockNumber),
+		zap.String("commit timestamp", txState.CommitTS.String()),
+		zap.Bool("paid", txState.Paid),
+		zap.String("keeper address", keeper.Address.String()),
+		zap.String("keeper day limit", keeper.DayLimit.String()),
+		zap.String("keeper day limit", keeper.SpentToday.String()),
+		zap.String("keeper last day ", keeper.LastDay.String()),
+		zap.Bool("keeper frozen", keeper.Frozen))
+
+	if keeper.Frozen {
```

Comment:
same as above: you logging about success but then failing with the required check.